buildPythonPackage: Enable direct attribute overriding via overrideAttrs (the non-staging ones)

ShamrockLee opened the pull request:
This PR makes the following attributes overridable using `overrideAttrs`:

- `src`
- `removePythonDeps`
- `relaxPythonDeps`
- `passthru.build-system`
- `passthru.dependencies`
- `passthru.pyproject`
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
python3Packages.buildPythonPackage: pass passthru.dependencies empty or not

ShamrockLee committed:
python3Packages.buildPythonPackage: make passthru.{dependencies,build-system} overridable

ShamrockLee committed:
python3Packages.buildPythonPackage: make passthru.pyproject overridable

ShamrockLee committed:
python3Packages.buildPythonPackage: make src overridable

ShamrockLee committed:
python3Packages.buildPythonPackage: use pythonRelaxDepsHook iff related attributes non-empty

ShamrockLee committed:
python3Packages.buildPythonPackage: remove ineffective outputs cleanAttrs

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-it-possible-to-override-cargosha256-in-buildrustpackage/4393/22

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5272

ShamrockLee commented:
I bumped the feature branch and edited the commit summaries.

ShamrockLee commented:
@ofborg build ansible ansible-later ansible-lint ansible-navigator ansible_2_16 ansible_2_17 ansible_2_18 molecule python312Packages.ansible-compat python312Packages.ansible-core python312Packages.ansible-kernel python312Packages.ansible-runner python312Packages.ansible-vault-rw python312Packages.molecule python312Packages.pytest-ansible python312Packages.pytest-testinfra scap-security-guide

ShamrockLee commented:
The package `ansible-later` fails to build on the `master` branch (see issue #388909). Let's try again without it.

@ofborg build ansible ansible-lint ansible-navigator ansible_2_16 ansible_2_17 ansible_2_18 molecule python312Packages.ansible-compat python312Packages.ansible-core python312Packages.ansible-kernel python312Packages.ansible-runner python312Packages.ansible-vault-rw python312Packages.molecule python312Packages.pytest-ansible python312Packages.pytest-testinfra scap-security-guide

ShamrockLee commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 376372 --skip-package-regex .*`

---
### `x86_64-linux`

---
### `aarch64-linux`

---
### `x86_64-darwin`

---
### `aarch64-darwin`

ShamrockLee commented:
Rebased to resolve merge conflicts after tree-wide formatting.

RossSmyth reviewed (approved):
This looks great.

ShamrockLee commented:
Pinging the Python people for inputs about the proposed `<pkg>.overrideAttrs` interface:
@hexa @mweinelt @natsukium

This PR will enable `<pkg>.overrideAttrs` overriding `src`, `removePythonDeps`, and `relaxPythonDeps`  directly; `build-system`, `dependencies`, and `pyproject` via `passthru`. Is it desirable?

ShamrockLee committed:
python3Packages.buildPythonPackage: add internal helper getFinalPassthru

Add getFinalPassthru, an internal helper function that
produces informative error message
when an atttribute is not found in `finalAttrs.passthru`.

ShamrockLee commented:
Add informative error message when `passthru` is emptied or overwritten during `<pkg>.overrideAttrs` overriding.
